src/serialize/root.zig

```
@@ -0,0 +1,133 @@
+            } else @panic("given type is not a simple type and cannot be serialized directly");
+        },
+        else => @panic("given type is not a simple type and cannot be serialized directly"),
+    }
+}
+
+fn serializeField(allocator: std.mem.Allocator, t: std.builtin.Type, key: []const u8, value: anytype, writer: anytype) !void {
+    switch (t) {
+        .int, .float, .bool => {
+            try writer.print("{s} = ", .{key});
+            try serializeSimpleValue(allocator, t, value, writer);
+        },
+        .pointer => {
+            try writer.print("{s} = ", .{key});
+            if (t.pointer.child == u8 and t.pointer.size == .Slice and t.pointer.is_const)
+                try serializeSimpleValue(allocator, t, value, writer);
+        },
+        .array => {
+            try writer.print("{s} = [ ", .{key});
+            if (t.array.len != 0) {
+                var i: usize = 0;
+                while (i < t.array.len - 1) {
+                    const elm = value[i];
+                    try serializeSimpleValue(allocator, @typeInfo(t.array.child), elm, writer);
+                    try writer.print(", ", .{});
+                    i += 1;
+                }
+            }
+            const elm = value[t.array.len - 1];
+            try serializeSimpleValue(allocator, @typeInfo(t.array.child), elm, writer);
+            try writer.print(" ]", .{});
+        },
+        .@"struct" => {
+            try writer.print("[{s}]\n", .{key});
+            try serializeStruct(allocator, value, writer);
+        },
+        else => {},
+    }
+}
+
+fn getFields(tinfo: std.builtin.Type) std.BoundedArray(std.builtin.Type.StructField, MAX_FIELD_COUNT) {
+    comptime var field_names = std.BoundedArray(std.builtin.Type.StructField, MAX_FIELD_COUNT).init(0) catch unreachable;
+    comptime var i: u8 = 0;
+    const fields = tinfo.@"struct".fields;
+    if (fields.len > MAX_FIELD_COUNT) @panic("struct field count exceeded MAX_FIELD_COUNT");
```

We should avoid `@panic` in libraries. We can return an error. The user can decide what to do in case of error.

---

Of-course I will deal with that. But as I said this is mostly preliminary work so it will take me some time to iron out things.